src/sdf/sdf-parse.c

```diff
@@ -0,0 +1,2739 @@
+
+   consume(tRPAREN);
+
+   return constr;
+}
+
+static sdf_node_t p_skew_constraint(void)
+{
+   // skew_constraint ::=
+   //       ( SKEWCONSTRAINT port_spec value )
+
+   BEGIN("skew constraint");
+
+   consume(tLPAREN);
+   consume(tSKEWCONSTR);
+
+   sdf_node_t constr = sdf_new(S_CONSTRAINT);
+   sdf_set_subkind(constr, S_CONSTR_KIND_SKEW);
+
+   sdf_add_signal(constr, p_port_spec());
+   sdf_add_value(constr, p_value());
+
+   consume(tRPAREN);
+
+   return constr;
+}
+
+static sdf_node_t p_arrival_env(void)
+{
+   // arrival_env ::=
+   //       ( ARRIVAL [ port_edge ] port_instance rvalue rvalue rvalue rvalue )
+
+   BEGIN("arrival env");
+
+   consume(tLPAREN);
+   consume(tARRIVAL);
+
+   sdf_node_t tenv = sdf_new(S_TIMING_ENV);
+   sdf_set_subkind(tenv, S_TENV_KIND_ARRIVAL);
+
+   if (scan(tLPAREN))
+      sdf_add_signal(tenv, p_port_edge());
+   sdf_add_signal(tenv, p_port_instance());
+
+   for (int i = 0; i < 4; i++)
+      sdf_add_value(tenv, p_rvalue());
+
+   consume(tRPAREN);
+
+   return tenv;
+}
+
+static sdf_node_t p_departure_env(void)
+{
+   // departure_env ::=
+   //       ( DEPARTURE [ port_edge ] port_instance rvalue rvalue rvalue rvalue )
+
+   BEGIN("departure env");
+
+   consume(tLPAREN);
+   consume(tDEPARTURE);
+
+   sdf_node_t tenv = sdf_new(S_TIMING_ENV);
+   sdf_set_subkind(tenv, S_TENV_KIND_DEPARTURE);
+
+   if (scan(tLPAREN))
+      sdf_add_signal(tenv, p_port_edge());
+   sdf_add_signal(tenv, p_port_instance());
+
+   for (int i = 0; i < 4; i++)
+      sdf_add_value(tenv, p_rvalue());
+
+   consume(tRPAREN);
+
+   return tenv;
+}
+
+static sdf_node_t p_slack_env(void)
+{
+   // slack_env ::=
+   //       ( SLACK port_instance rvalue rvalue rvalue rvalue [ real_number ] )
+
+   BEGIN("slack env");
+
+   consume(tLPAREN);
+   consume(tSLACK);
+
+   sdf_node_t tenv = sdf_new(S_TIMING_ENV);
+   sdf_set_subkind(tenv, S_TENV_KIND_SLACK);
+
+   sdf_add_signal(tenv, p_port_instance());
+
+   for (int i = 0; i < 4; i++)
+      sdf_add_value(tenv, p_rvalue());
+
+   if (scan(tREAL, tINT))
+      sdf_add_value(tenv, p_real_number(true));
+
+   consume(tRPAREN);
+
+   return tenv;
+}
+
+static sdf_node_t p_waveform_env(void)
+{
+   // waveform_env ::=
+   //       ( WAVEFORM port_instance real_number edge_list )
+
+   BEGIN("waveform env");
+
+   consume(tLPAREN);
+   consume(tWAVEFORM);
+
+   sdf_node_t tenv = sdf_new(S_TIMING_ENV);
+   sdf_set_subkind(tenv, S_TENV_KIND_WAVEFORM);
+
+   sdf_add_signal(tenv, p_port_instance());
+   sdf_add_value(tenv, p_real_number(true));
+
+   // sdf_value of S_TENV_KIND_WAVEFORM
+   //    0     - real_number
+   //    1 - X - Edge pairs
+
+   p_edge_list(tenv);
+
+   consume(tRPAREN);
+
+   return tenv;
+}
+
+static void p_te_spec(sdf_node_t cell)
+{
+   // te_spec ::=
+   //       ( TIMINGENV te_def { te_def } )
+
+   BEGIN("te spec");
+
+   consume(tLPAREN);
+   consume(tTIMINGENV);
+
+   // TODO: te_def should be obligatory at least once
+   while (scan(tLPAREN)) {
```

Comment:
This can be done by replacing the while loop with do-while (same in other places).